tools/coq_dune.ml

```diff
@@ -0,0 +1,245 @@
+  let ml4_rule = "(run coqp5 -loc loc -impl %{pp-file} -o %{targets})" in
+  pp_rule fmt [target] [ml] ml4_rule
+
+let pp_mlg_dep _dir fmt ml =
+  let target = Filename.(remove_extension ml) ^ ".ml" in
+  let ml4_rule = "(run coqpp %{pp-file})" in
+  pp_rule fmt [target] [ml] ml4_rule
+
+let pp_dep dir fmt oo = match oo with
+  | VO vo -> pp_vo_dep dir fmt vo
+  | ML4 f -> pp_ml4_dep dir fmt f
+  | MLG f -> pp_mlg_dep dir fmt f
+
+let out_install fmt dir ff =
+  let itarget = String.concat "/" dir in
+  let ff = pmap (function | VO vo -> Some vo.target | _ -> None) ff in
+  let pp_ispec fmt tg = fprintf fmt "(%s as %s)" tg (itarget^"/"^tg) in
+  fprintf fmt "(install@\n @[(section lib)@\n(files @[%a@])@])@\n"
+    (pp_list pp_ispec sep) ff
+
+(* For each directory, we must record two things, the build rules and
+   the install specification. *)
+let record_dune d ff =
+  let sd = String.concat "/" d in
+  if Sys.file_exists sd && Sys.is_directory sd then
+    let out = open_out (sd^"/dune") in
+    let fmt = formatter_of_out_channel out in
+    if List.nth d 0 = "plugins" then
+      fprintf fmt "(include plugin_base.dune)@\n";
+    out_install fmt d ff;
+    List.iter (pp_dep d fmt) ff;
+    fprintf fmt "%!";
+    close_out out
+  else
+    eprintf "error in coq_dune, a directory disappeared: %s@\n%!" sd
+
+(*  *)
+let choose_ml4g_form f =
+  if Filename.check_suffix f ".ml4" then ML4 f
+  else MLG f
+
+let scan_mlg4 m d =
+  let dir = ["plugins"; d] in
+  let m = DirMap.add dir [] m in
+  let ml4 = Sys.(List.filter (fun f -> Filename.(check_suffix f ".ml4" || check_suffix f ".mlg"))
+                   Array.(to_list @@ readdir String.(concat "/" dir))) in
+  List.fold_left (fun m f -> add_map_list ["plugins"; d] (choose_ml4g_form f) m) m ml4
+
+let scan_plugins m =
+  let dirs = Sys.(List.filter (fun f -> is_directory @@ "plugins/"^f) Array.(to_list @@ readdir "plugins/")) in
+  List.fold_left scan_mlg4 m dirs
+
+(* Process .vfiles.d and generate a skeleton for the dune file *)
+let parse_coqdep_line l =
+  match Str.(split (regexp ":") l) with
+  | [targets;deps] ->
+    let targets = Str.(split (regexp "[ \t]+") targets) in
+    let deps    = Str.(split (regexp "[ \t]+") deps) in
+    let targets = filter_no_vo targets in
+    begin match targets with
+    | [target] ->
+      let dir, target = Filename.(dirname target, basename target) in
+      Some (String.split_on_char '/' dir, VO { target; deps; })
+    (* Otherwise a vio file, we ignore *)
+    | _ -> None
+    end
+  (* Strange rule, we ignore *)
+  | _ -> None
+
+let rec read_vfiles ic map =
+  try
+    let rule = parse_coqdep_line (input_line ic) in
+    (* Add vo_entry to its corresponding map entry *)
+    let map = option_cata map (fun (dir, vo) -> add_map_list dir vo map) rule in
+    read_vfiles ic map
+  with End_of_file -> map
+
+let out_map map =
+  DirMap.iter record_dune map
+
+let exec_ifile f =
+  match Array.length Sys.argv with
```

Comment:
You can use OCaml's syntax to match on an array and have the compiler emit roughly that code for you. Eg
```ocaml
match Sys.argv with
| [| |] -> f stdin
| [| file |] -> let ic = open_in ..
```